### PR TITLE
Change regex flags and add unit tests for word boundaries

### PR DESCRIPTION
**Pull request by jgutman**
### Intent

Addresses both #9344 and #9813. 

### Approach

Apparently, `grep` flavor regexes within the `boost::regex_replace` function don't process `\b` as word boundaries, even though grep on the command line, which we use to find the matches in the first place, works just fine. [This page](https://www.boost.org/doc/libs/1_76_0/libs/regex/doc/html/boost_regex/syntax/basic_syntax.html) indicates that the `egrep` flavor which supports POSIX ERE, and is equivalent to `grep -E` in the command line, resolves the original issue.

By adding support for ERE, we have now changed which characters do and do not need to be escaped. In general, only escape sequences such as `\b` need to be escaped.  Characters with special regex meaning

```
.[{}()\*+?|^$
```
will now be treated as literals if escaped with a backslash, and will retain their special regex meaning if unescaped.

- Quantifiers (such as `?` and `+`) were previously supported but were treated as literals if unescaped, and needed to be escaped with `\` to be treated as quantifiers. They will now be treated as quantifiers if unescaped, and will need to be escaped to be treated as literals.
- Parentheses and curly braces `()` and `{}` were previously supported but were treated as literals if unescaped, and needed to be escaped with to be treated as their special regex meaning (e.g. marked sub-expressions, bounded repeats). They will now be parsed with their regex meaning if unescaped, and will need to be escaped to be treated as literals.
- The same is true for other characters such as `|`, `^`, and `$` which previously needed to be escaped and were treated as literals if unescaped. They now do not need to be escaped and will be treated as literals only if escaped.
- Extended regex escape sequences, such as `\w`, `\W`, `\b`, `\B`, `\d`, `\D` that are part of the POSIX ERE standard, are now supported, during find, preview, and replace. See [this page](https://www.boost.org/doc/libs/1_71_0/libs/regex/doc/html/boost_regex/syntax/basic_extended.html#boost_regex.syntax.basic_extended.single_character_character_class) for more examples

See also [this](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX-Extended_Regular_Expressions#Table_of_metacharacters) for a table of the differences in escaping between BRE and ERE.

### Automated Tests

Existing CPP unit tests in [SessionFindTests.cpp](https://github.com/rstudio/rstudio/blob/12bc87d051519f11c096024023cc8996616b138f/src/cpp/session/modules/SessionFindTests.cpp) all pass. I've also added two new test cases to specifically test the word boundary handling

### QA Notes

Should be QA'd on at least Windows and either Mac or Linux, since the find part of Find in Files is handled slightly differently on Windows. 

Should test that when "Whole Word" is checked under Find in Files, the find matches, preview replacements, and final replacements match whole word matches only and do not replace partial word matches (e.g. `map` replaces the `map` in `map this!` and `map(1:5)` but not `mapping`)

Should test that when Regular expression is checked, and special characters such as `.[{}()\*+?|^$` are escaped with a single backslash, they are matched literally.

Should test that when Regular expression is checked and and special characters such as `.[{}()\*+?|^$` are not escaped with a single backslash, they are treated as special regex characters (e.g. `+` means match the preceding item 1 or more times, `{n}` means match the preceding item exactly n times, etc, using the POSIX extended regular expression standard.

Should test that escape sequences such as `\b` (word boundary), `\w` (word-like character) `\d` (digit), `\s` (whitespace characters) and their inverses (e.g. `\B` `\W` `\D`) match expected characters.

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [X] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [X] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [X] This PR passes all local unit tests

### CODE REVIEW

**Comment by kevinushey** [comment]
I think rather than `emacs` we want to just use the default (`perl`). See e.g. https://www.boost.org/doc/libs/1_50_0/libs/regex/doc/html/boost_regex/syntax/perl_syntax.html

My guess is that we chose to use `boost::regex::grep` here because we wanted to be sure the regular expression "flavor" used by the `grep` utility matched what we were trying to use in Boost. I think we wanted to use PCREs with `grep`, but the version of `grep` bundled on Windows is too old:

https://github.com/rstudio/rstudio/issues/3572

So, in order to "fully" fix this I think we should also explore updating the bundled copy of grep we use on Windows.

**Comment by jgutman** [comment]
Yeah, that makes sense to me, this still doesn't give us matching behaviors between the find and replace for things like `\+` and `\?` either so I don't particularly love the solution

**Comment by kevinushey** [comment]
It also looks like we aren't actually trying to request extended regular expressions (or perl regular expressions) here:

https://github.com/rstudio/rstudio/blob/01ef6e05e3a5e02f39741e78cc6519b6b074b20d/src/cpp/session/modules/SessionFind.cpp#L1433-L1454

I wonder if adding `-E` or `-P` (and also updating our usages of `boost::regex` in this PR) would suffice?

File: src/cpp/session/modules/SessionFind.cpp [hunk 1]
```diff
@@ -1404,7 +1404,8 @@ core::Error runGrepOperation(const GrepOptions& grepOptions, const ReplaceOption
       cmd << "-c" << "grep.fullName=false";
       cmd << "-C";
       cmd << string_utils::utf8ToSystem(dirPath.getAbsolutePath());
-      cmd <<  "grep";
+      cmd << "grep";
+      cmd << "-E"; // use extended-grep (egrep) for Extended Regular Expressions
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionFind.cpp:1408]: We should only set this when `grepOptions.asRegex()` is true.

File: src/cpp/session/modules/SessionFind.cpp [hunk 2]
```diff
@@ -1431,6 +1432,7 @@ core::Error runGrepOperation(const GrepOptions& grepOptions, const ReplaceOption
    }
    else
    {
+      cmd << "-E"; // use extended-grep (egrep) for Extended Regular Expressions
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionFind.cpp:1435]: Same as above.

Comment (kevinushey) [reply src/cpp/session/modules/SessionFind.cpp:1435]: (You might want to combine with https://github.com/rstudio/rstudio/pull/10362/files?w=1#diff-94062616737b3005a9fa89d60aa0596694818c97563995b50cba4661956d69dbR1448-R1449)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/output/find/FindOutputPresenter.java [hunk 3]
```diff
@@ -302,8 +302,12 @@ public void execute()
                         for (String pattern : dialogState_.getExcludeFilePatterns())
                            excludeFilePatterns.push(pattern);
 
-                        server_.completeReplace(dialogState_.getQuery(),
-                                                dialogState_.isRegex(),
+                        String serverQuery = dialogState_.getQuery();
```
Comment (jgutman) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/output/find/FindOutputPresenter.java:305]: Previously, we were padding with `\b` in the find operation but not in the complete replace operation, which is why the find and preview respected the whole word option, but complete replace did not. This fixes that issue. 

This bit could have been separated out into its own PR, since it was strictly a failure to pass the correct args to the RPC, but for purposes of testing I've included it together with the other changes.

Comment (kevinushey) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/output/find/FindOutputPresenter.java:305]: Nice catch!

File: src/cpp/session/modules/SessionFindTests.cpp [hunk 4]
```diff
@@ -42,11 +42,16 @@ const size_t rMatchOff = 10;
 const size_t caseMatchOn = 21;
 const size_t caseMatchOff = 27;
 
-const std::string kFindRegex("\\([a-z]\\)\\1\\{2\\}\\([a-z]\\)\\2\\{2\\}");
+const std::string kFindRegex("([a-z])\\1{2}([a-z])\\2{2}");
```
Comment (melissa-barca) [inline src/cpp/session/modules/SessionFindTests.cpp:45]: Were the original escaping patterns here not actually needed or do your changes affect how the user needs to enter their queries? If the ladder, can you include the differences in the description of this PR?

File: src/cpp/session/modules/SessionFindTests.cpp [hunk 5]
```diff
@@ -246,6 +275,36 @@ TEST_CASE("SessionFind")
...
+      const size_t matchOn = 2;
+      const size_t matchOff = 7;
+      size_t replaceMatchOff;
+
+      replacer.replaceRegex(matchOn, matchOff, regex, replacement, &line, &replaceMatchOff);
+      CHECK(line.compare("heLO!") == 0);
+      CHECK(replaceMatchOff == 4);
+   }
+
+   SECTION("Replace regex with quantifiers")
```
Comment (melissa-barca) [inline src/cpp/session/modules/SessionFindTests.cpp:293]:
```suggestion
   SECTION("Replace regex with quantifiers and backreference")
```
We should use unique section names for these tests so that it's obvious which test is failing when there is an issue.

**Review by melissa-barca — CHANGES_REQUESTED** [review]
Really awesome job with this! Could you list in the PR what functionality we are gaining with the switch to extended regexes? We should add unit test coverage for any new functionality. 
Could you fill in the QA notes section as well? It will be very important to test this thoroughly on all OSes (windows vs mac vs linux) as well as any new functionality or changes to the way the user escapes queries/replace text. It'd may be helpful to include the section title list of unit tests we have coverage for under the "Automated Tests" section.

File: src/cpp/session/modules/SessionFindTests.cpp (hunk 4 above)
Comment (jgutman) [reply src/cpp/session/modules/SessionFindTests.cpp:45]: Yes, I think that with ERE enabled, and the regex option selected, the user does not need to escape parentheses and curly braces. If they want a literal parenthesis or curly brace, or if they want a literal `?` or `+`, they will now escape it. Will do!

File: src/cpp/session/modules/SessionFind.cpp [hunk 6]
```diff
@@ -1753,7 +1757,11 @@ core::Error Replacer::replaceRegexIgnoreCase(size_t matchOn, size_t matchOff,
 {
    try
    {
-      boost::regex find(findRegex, boost::regex::grep | boost::regex::icase);
+
+      // egrep (grep -E) flavor regex includes POSIX ERE
+      boost::regex_constants::syntax_option_type flags = boost::regex::egrep & boost::regex::bk_vbar;
+      flags = flags | boost::regex::icase;
```
Comment (jgutman) [inline src/cpp/session/modules/SessionFind.cpp:1763]: I wasn't 100% sure about the difference between `&` and `|` in terms of these flags, how they're combined, and their precedence, but this seemed in line with other usages I saw for these flags

**Review by melissa-barca — APPROVED** [review]
LGTM!

File: src/cpp/session/modules/SessionFind.cpp (hunk 6 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionFind.cpp:1763]: These are bitflags; those are usually combined with `|`. `&` is usually used to check if a particular bitflag is "active" in a set.

Comment (melissa-barca) [reply src/cpp/session/modules/SessionFind.cpp:1763]: Hm, using `&` here means we're not actually passing either flag to our regex and are relying on defaults (perl regexes). You'd get the same result if you only set `boost::regex::icase`.

Comment (kevinushey) [reply src/cpp/session/modules/SessionFind.cpp:1763]: That is surprising to me. You can see how flags are defined and combined here, for example, in the boost.regex source code:

https://github.com/boostorg/regex/blob/a9f18c2c03a5e64745ec8fcdb8247dbb2a84e9df/include/boost/regex/v4/regbase.hpp#L93-L104

I wonder if we want to start with "normal" and then add some flags to turn off PCRE-specific features? (Or, just use `perl` anyway?)

**Comment by jgutman** [comment]
I think the reason we decided not to use perl was that we'd then need grep
-P for the find to truly be aligned between the find and the replace, and
grep -P is not available on Mac system's grep. But maybe that's not
actually a concern. I guess I'm not sure then exactly what's failing when I
use `flags = boost::regex::egrep | boost::regex::bk_vbar;`. The original
suggestion I saw was to use `flags = boost::regex::egrep  &
~boost::regex::no_bk_refs;` but I'm not sure if that's correct

File: src/cpp/session/modules/SessionFind.cpp (hunk 6 above)
Comment (melissa-barca) [reply src/cpp/session/modules/SessionFind.cpp:1763]: Relying on PCRE seems reasonable to me, especially since either way we were announcing a change in behavior and all of our unit tests pass on all platforms. Is that what R uses?

Comment (jgutman) [reply src/cpp/session/modules/SessionFind.cpp:1763]: I think not by default -- R's grep functions all take a `perl` argument, but it's FALSE by default

**Comment by melissa-barca** [comment]
The tilde is bitwise NOT so that suggestion is equivalent to `flags = boost::regex::egrep;`

**Comment by jgutman** [comment]
> The tilde is bitwise NOT so that suggestion is equivalent to `flags = boost::regex::egrep;`

No but I think it's a double negative, ie `boost::regex::bk_vbar` (enable backrefernces) and `~boost::regex::no_bk_refs` (turn off disabling backreferences) should be the same. Either way, I don't seem to be able to combine them with `egrep` using an `|` and I'm not clear on why

**Comment by kevinushey** [comment]
I think all (or at least most) extended regular expressions are also valid perl regular expressions, so it might be simplest to just use `normal` (or `perl`) for our Boost regular expressions.

**Comment by melissa-barca** [comment]
Ah - I didn't realize the double negative/variable name change, thank you.

From boost's regbase.hpp it looks like no_bk_refs in specific to perl and you'll notice the `egrep` is derived from `extended` which specifies `no_bk_refs`. 

```
   58       // options specific to perl group:
   59       //
   60       no_bk_refs = 1 << 8,                        // \d not allowed
   61       no_perl_ex = 1 << 9,                        // disable perl extensions
   62       no_mod_m = 1 << 10,                         // disable Perl m modifier
   63       mod_x = 1 << 11,                            // Perl x modifier
   64       mod_s = 1 << 12,                            // force s modifier on (overrides match_not_dot_newline)
   65       |o_mod_s = 1 << 13,                         // force s modifier off (overrides match_not_dot_newline)
   66
   67       //
   68       // options specific to basic group:
   69       //
   70       no_char_classes = 1 << 8,                   // [[:CLASS:]] not allowed
   71       no_intervals = 1 << 9,                      // {x,y} not allowed
   72       bk_plus_qm = 1 << 10,                       // uses \+ and \?
   73       bk_vbar = 1 << 11,                          // use \| for alternatives
   74       emacs_ex = 1 << 12,                         // enables emacs extensions
   75
   76       //
   77       // options common to all groups:
   78       //
   79       no_escape_in_lists = 1 << 16,                     // '\' not special inside [...]
   80       newline_alt = 1 << 17,                            // \n is the same as |
   81       no_except = 1 << 18,                              // no exception on error
   82       failbit = 1 << 19,                                // error flag
   83       icase = 1 << 20,                                  // characters are matched regardless of case
   84       nocollate = 0,                                    // don't use locale specific collation (deprecated)
   85       collate = 1 << 21,                                // use locale specific collation
   86       nosubs = 1 << 22,                                 // don't mark sub-expressions
   87       save_subexpression_location = 1 << 23,            // save subexpression locations
   88       no_empty_expressions = 1 << 24,                   // no empty expressions allowed
   89       optimize = 0,                                     // not really supported
   90
   91
   92
   93       basic = basic_syntax_group | collate | no_escape_in_lists,
   94       extended = no_bk_refs | collate | no_perl_ex | no_escape_in_lists,
   95       normal = 0,
   96       emacs = basic_syntax_group | collate | emacs_ex | bk_vbar,
   97       awk = no_bk_refs | collate | no_perl_ex,
   98       grep = basic | newline_alt,
   99       egrep = extended | newline_alt,
```

It might be worth further investigating if there are any scenarios where find doesn't align with replace in the current scenario.

**Comment by melissa-barca** [comment]
> I think all (or at least most) extended regular expressions are also valid perl regular expressions, so it might be simplest to just use `normal` (or `perl`) for our Boost regular expressions.

Yes, as far as I can tell ERE are a subset of PCRE so searching with ERE and replacing with PCRE shouldn't be a problem (while searching with PCRE and replacing with ERE would cause problems).
